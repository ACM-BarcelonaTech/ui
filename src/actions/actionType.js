
const CELL_SETS = 'CELL_SETS';
const LOAD_CELL_SETS = `${CELL_SETS}.LOAD`;
const UPDATE_CELL_SETS = `${CELL_SETS}.UPDATE`;
const CELL_SETS_COLOR = `${CELL_SETS}.COLOR`;

const CELLS = 'CELLS';
const LOAD_CELLS = `${CELLS}.LOAD`;

// eslint-disable-next-line import/prefer-default-export
export {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, LOAD_CELLS, CELL_SETS_COLOR,
};
