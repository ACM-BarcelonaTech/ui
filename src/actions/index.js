import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, CREATE_CLUSTER, LOAD_CELLS, CELL_SETS_COLOR,
} from './actionType';
import { connectionPromise } from '../components/content-wrapper/ContentWrapper';


// eslint-disable-next-line func-names
const loadCellSets = (experimentId) => function (dispatch, getState) {
  if (getState().cellSets.data) {
    return Promise.resolve();
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      data: json.cellSets,
    }),
  ).catch((e) => console.log('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (newState) => function (dispatch, getState) {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  return dispatch({
    type: UPDATE_CELL_SETS,
    data: newState,
  });
};

const createCluster = (cellSetInfo, clusterName, color) => function (dispatch, getState) {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  return dispatch({
    type: CREATE_CLUSTER,
    data: newCluster,
  });
};

// eslint-disable-next-line func-names
const loadCells = (experimentId, requestBody) => function (dispatch, getState) {
  if (getState().cells.data) {
    return Promise.resolve();
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body: requestBody,
    };

    io.emit('WorkRequest', request);

    console.log('emitted!!!');

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      console.log('response! ');
      return dispatch({
        type: LOAD_CELLS,
        data: embedding,
      });
    });
  });
};

// eslint-disable-next-line func-names
const cellSetsColor = (colorData) => function (dispatch, getState) {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return Promise.resolve();
};

// eslint-disable-next-line import/prefer-default-export
export {
  loadCellSets, updateCellSets, createCluster, loadCells, cellSetsColor,
};
