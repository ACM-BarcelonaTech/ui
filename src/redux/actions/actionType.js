
const CELL_SETS = 'CELL_SETS';
const LOAD_CELL_SETS = `${CELL_SETS}.LOAD`;
const UPDATE_CELL_SETS = `${CELL_SETS}.UPDATE`;
const CREATE_CLUSTER = `${CELL_SETS}.CREATE`;
const CELL_SETS_COLOR = `${CELL_SETS}.COLOR`;

const CELLS = 'CELLS';
const LOAD_CELLS = `${CELLS}.LOAD`;

const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

export {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, CREATE_CLUSTER, LOAD_CELLS, CELL_SETS_COLOR,
  LOAD_GENE_LIST, UPDATE_GENE_LIST,
};
