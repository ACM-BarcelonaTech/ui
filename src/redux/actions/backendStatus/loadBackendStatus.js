import handleError from 'utils/http/handleError';
import fetchAPI from 'utils/http/fetchAPI';
import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_ERROR,
} from 'redux/actionTypes/backendStatus';

import config from 'config';
import { api } from 'utils/constants';
import endUserMessages from 'utils/endUserMessages';

const loadBackendStatus = (experimentId) => async (dispatch) => {
  dispatch({
    type: BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  let url;

  if (config.currentApiVersion === api.V1) {
    url = `/v1/experiments/${experimentId}/backendStatus`;
  } else if (config.currentApiVersion === api.V2) {
    url = `/v2/experiments/${experimentId}/backendStatus`;
  }

  try {
    const status = await fetchAPI(url);

    dispatch({
      type: BACKEND_STATUS_LOADED,
      payload: {
        experimentId,
        status,
      },
    });

    return status;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_BACKEND_STATUS);

    dispatch({
      type: BACKEND_STATUS_ERROR,
      payload: {
        experimentId,
        error: errorMessage,
      },
    });
  }
};

export default loadBackendStatus;
