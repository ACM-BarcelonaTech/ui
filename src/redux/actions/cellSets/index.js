import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import resetCellSets from './resetCellSets';

import loadCellSets from './loadCellSets';
import saveCellSets from './saveCellSets';
import updateCellSetHierarchy from './updateCellSetHierarchy';
import updateCellSetProperty from './updateCellSetProperty';
import updateCellSetSelected from './updateCellSetSelected';

export {
  createCellSet,
  deleteCellSet,
  resetCellSets,
  loadCellSets,
  saveCellSets,
  updateCellSetHierarchy,
  updateCellSetSelected,
  updateCellSetProperty,
};
