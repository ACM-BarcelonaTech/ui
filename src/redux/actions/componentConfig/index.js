import loadPlotConfig from './loadPlotConfig';
import loadComponentConfig from './loadComponentConfig';
import updatePlotConfig from './updatePlotConfig';
import savePlotConfig from './savePlotConfig';
import fetchPlotDataWork from './fetchPlotDataWork';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  fetchPlotDataWork,
};
