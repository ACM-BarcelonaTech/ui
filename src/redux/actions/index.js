/* eslint-disable no-param-reassign */
import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, PUSH_CELL_SETS, CREATE_CLUSTER, CELL_SETS_COLOR,
  UPDATE_GENE_LIST, LOAD_GENE_LIST, SELECTED_GENES, UPDATE_GENE_EXPRESSION,
  LOAD_CELLS, BUILD_HEATMAP_SPEC, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from './actionType';
import connectionPromise from '../../utils/socketConnection';

const loadCellSets = (experimentId) => (dispatch, getState) => {
  if (getState().cellSets.data) {
    return null;
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      experimentId,
      data: json.cellSets,
    }),
  ).catch((e) => console.error('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (experimentId, newState) => (dispatch, getState) => {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  dispatch({
    type: UPDATE_CELL_SETS,
    experimentId,
    data: newState,
  });

  return dispatch(pushCellSets(experimentId));
};

const pushCellSets = (experimentId) => (dispatch, getState) => {
  // If no loaded data exists for our cell sets and some event would
  // trigger a push, we do not want to overwrite our cell sets with
  // empty data.
  if (!getState()?.cellSets?.data) {
    return null;
  }

  return fetch(
    `${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(
        getState().cellSets.data,
        (k, v) => ((k === 'title') ? undefined : v),
      ),
    },
  ).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: PUSH_CELL_SETS,
      experimentId,
      data: json,
    }),
  ).catch((e) => console.error('Error when trying to push cell sets to API: ', e));
};

const createCluster = (experimentId, cellSetInfo, clusterName, color) => (dispatch) => {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  dispatch({
    type: CREATE_CLUSTER,
    experimentId,
    data: newCluster,
  });

  return dispatch(pushCellSets(experimentId));
};

const cellSetsColor = (colorData) => (dispatch, getState) => {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return null;
};

const loadCells = (experimentId, embeddingType) => (dispatch, getState) => {
  if (getState().cells.data) {
    return null;
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const body = {
      name: 'GetEmbedding',
      type: embeddingType,
    };

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      return dispatch({
        type: LOAD_CELLS,
        experimentId,
        data: embedding,
      });
    });
  });
};

const updateGeneList = (experimentId, tableState) => (dispatch, getState) => {
  if (getState().geneList?.tableState === tableState) {
    return null;
  }
  dispatch({
    type: LOAD_GENE_LIST,
  });
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();
    const { geneList } = getState();

    const orderBy = tableState?.sorter.field
      || geneList.tableState?.sorter.field
      || 'dispersions';

    const orderDirection = tableState?.sorter.order
      || geneList.tableState?.sorter.order
      || 'ascend';

    const currentPage = tableState?.pagination.current
      || geneList.tableState?.pagination.current
      || 1;

    const currentPageSize = tableState?.pagination.pageSize
      || geneList.tableState?.pagination.pageSize
      || 1;

    const body = {
      name: 'ListGenes',
      selectFields: ['gene_names', 'dispersions'],
      orderBy,
      orderDirection: (orderDirection === 'ascend') ? 'ASC' : 'DESC',
      offset: ((currentPage - 1) * currentPageSize),
      limit: currentPageSize,
    };

    if (tableState.geneNamesFilter) {
      body.geneNamesFilter = tableState.geneNamesFilter;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);
    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const data = JSON.parse(res.results[0].body);

      const { total, rows } = data;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      if (tableState && tableState.pagination) {
        tableState.pagination.total = total;
      }
      return dispatch({
        type: UPDATE_GENE_LIST,
        experimentId,
        data: {
          rows,
          tableState: tableState || geneList.tableState,
        },
      });
    });
  });
};

const loadDiffExpr = (
  experimentId, comparisonType, firstSelectedCluster, secondSelectedCluster,
) => (dispatch) => {
  dispatch({
    type: LOAD_DIFF_EXPR,
  });
  return connectionPromise().then((io) => {
    const ComparisonTypes = {
      One: 'Versus Rest',
      Two: 'Across Sets',
    };

    const requestUuid = uuidv4();

    const body = {
      name: 'DifferentialExpression',
      maxNum: 100,
      cellSet: firstSelectedCluster.key,
    };
    if (comparisonType === ComparisonTypes.One) {
      body.compareWith = 'rest';
    } else {
      body.compareWith = secondSelectedCluster.key;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      let data = {};

      try {
        data = JSON.parse(res.results[0].body);
      } catch (error) {
        console.error(error);
        data = { rows: [] };
      }

      const { rows } = data;
      const total = rows.length;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      return dispatch({
        type: UPDATE_DIFF_EXPR,
        experimentId,
        data: {
          allData: rows,
          total,
        },
      });
    });
  });
};

const updateSelectedGenes = (genes, selected) => (dispatch, getState) => {
  const { selectedGenes, geneExperessionData } = getState();
  selectedGenes.geneList = selectedGenes.geneList || {};
  let newGenesAdded = false;
  genes.forEach((gene) => {
    if (selected) {
      if (!selectedGenes.geneList[gene]) {
        // This will present a dict of newly selected genes from gene table
        // { <Gene Name>: <Is the gene experession resolved?> }
        selectedGenes.geneList[gene] = false;
        newGenesAdded = true;
      }
    } else {
      delete selectedGenes.geneList[gene];
      const foundGene = geneExperessionData.data.findIndex((g) => g.geneName === gene);
      geneExperessionData.data.splice(foundGene, 1);
    }
  });
  dispatch({
    type: SELECTED_GENES,
    data: { newGenesAdded },
  });
  dispatch({
    type: UPDATE_HEATMAP_SPEC,
    data: {
      genes: geneExperessionData.data,
      rendering: true,
    },
  });
  setTimeout(() => {
    // Running this action with a delay to allow rerendering of heatmap
    dispatch({
      type: UPDATE_HEATMAP_SPEC,
      data: { rendering: false },
    });
  }, 50);
};

const loadGeneExpression = (experimentId) => (dispatch, getState) => {
  const { geneExperessionData } = getState();
  if (geneExperessionData?.isLoading) {
    return null;
  }
  const { newGenesAdded, geneList } = getState().selectedGenes;
  const newGeneBatch = [];
  if (newGenesAdded) {
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(geneList)) {
      if (!value) {
        newGeneBatch.push(key);
        geneList[key] = true;
      }
    }
  }
  dispatch({
    type: SELECTED_GENES,
    data: {
      newGenesAdded: false,
    },
  });
  if (newGeneBatch.length) {
    dispatch({
      type: UPDATE_GENE_EXPRESSION,
      data: {
        isLoading: true,
      },
    });
    return connectionPromise().then((io) => {
      const requestUuid = uuidv4();

      const body = {
        name: 'GeneExpression',
        cellSets: 'all',
        genes: newGeneBatch,
      };

      const request = {
        uuid: requestUuid,
        socketId: io.id,
        experimentId,
        timeout: '2021-01-01T00:00:00Z',
        body,
      };

      io.emit('WorkRequest', request);

      io.on(`WorkResponse-${requestUuid}`, (res) => {
        const heatMapData = JSON.parse(res.results[0].body);
        const { data } = getState().geneExperessionData;
        if (data) {
          Array.prototype.push.apply(heatMapData.data, data);
        }
        dispatch({
          type: UPDATE_GENE_EXPRESSION,
          data: {
            heatMapData,
            isLoading: false,
          },
        });
        return dispatch({
          type: BUILD_HEATMAP_SPEC,
          data: {
            geneExperessionData: heatMapData,
          },
        });
      });
    });
  }
};

export {
  loadCellSets,
  updateCellSets,
  pushCellSets,
  createCluster,
  loadCells,
  cellSetsColor,
  updateGeneList,
  updateSelectedGenes,
  loadGeneExpression,
  loadDiffExpr,
};
