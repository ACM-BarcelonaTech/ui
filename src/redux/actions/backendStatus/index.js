import loadBackendStatus from './loadBackendStatus';
import updateBackendStatus from './updateBackendStatus';

export {
  loadBackendStatus,
  updateBackendStatus,
};
