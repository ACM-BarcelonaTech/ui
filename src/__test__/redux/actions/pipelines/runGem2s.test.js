import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import initialExperimentsState, { experimentTemplate } from '../../../../redux/reducers/experiments/initialState';
import loadBackendStatus from '../../../../redux/actions/backendStatus/loadBackendStatus';
import initialBackendState from '../../../../redux/reducers/backendStatus';

import {
  EXPERIMENT_SETTINGS_PIPELINE_START,
  EXPERIMENT_SETTINGS_INFO_UPDATE,
} from '../../../../redux/actionTypes/experimentSettings';

import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_ERROR,
} from '../../../../redux/actionTypes/backendStatus';

import { runGem2s } from '../../../../redux/actions/pipeline';

const mockStore = configureStore([thunk]);
enableFetchMocks();

jest.mock('../../../../redux/actions/backendStatus/loadBackendStatus',
  () => jest.fn().mockImplementation(() => async () => { }));

const experimentId = 'experiment-id';
const projectId = 'project-id';
const oldGem2sHash = 'old-gem2s-hash';

const initialState = {
  experiments: {
    ...initialExperimentsState,
    [experimentId]: {
      ...experimentTemplate,
      name: 'Mock experiment',
      id: experimentId,
      projectUuid: projectId,
      sampleIds: ['sample-1', 'sample-2'],
    },
  },
  backendStatus: {
    [experimentId]: {
      ...initialBackendState,
      status: {
        gem2s: {
          paramsHash: oldGem2sHash,
        },
      },
    },
  },
};

describe('runGem2s action', () => {
  beforeEach(() => {
    const response = new Response(JSON.stringify({}));

    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockResolvedValueOnce(response);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Dispatches events properly', async () => {
    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(BACKEND_STATUS_LOADING);
    expect(actions[1].type).toEqual(EXPERIMENT_SETTINGS_PIPELINE_START);
    expect(loadBackendStatus).toHaveBeenCalled();

    expect(actions[2].type).toEqual(EXPERIMENT_SETTINGS_INFO_UPDATE);
    expect(actions).toMatchSnapshot();
  });

  it('Dispatches status error if loading fails', async () => {
    fetchMock.resetMocks();
    fetchMock.mockResponse(JSON.stringify({ message: 'some weird error that happened' }), { status: 400 });

    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    const actions = store.getActions();

    expect(actions[0].type).toEqual(BACKEND_STATUS_LOADING);
    expect(loadBackendStatus).not.toHaveBeenCalled();
    expect(actions[1].type).toEqual(BACKEND_STATUS_ERROR);

    expect(actions).toMatchSnapshot();
  });

  it('Should run GEM2S if hash is not defined', async () => {
    fetchMock.resetMocks();

    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    const requestBody = fetchMock.mock.calls[0][1].body;

    const { shouldRun } = JSON.parse(requestBody);

    expect(shouldRun).toEqual(true);
  });

  it('Should rerun GEM2S if passed in hash is different from old hash', async () => {
    fetchMock.resetMocks();

    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId, 'new-different-hash'));

    const requestBody = fetchMock.mock.calls[0][1].body;

    const expectedBody = JSON.stringify({
      shouldRun: true,
      gem2sHash: 'new-different-hash',
    });

    expect(requestBody).toEqual(expectedBody);
  });

  it('Should not run GEM2S if calculated hash is the same as old hash', async () => {
    fetchMock.resetMocks();

    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId, oldGem2sHash));

    const requestBody = fetchMock.mock.calls[0][1].body;

    const expectedBody = JSON.stringify({
      shouldRun: false,
      gem2sHash: oldGem2sHash,
    });

    expect(requestBody).toEqual(expectedBody);
  });

  it('Dispatches properly without project data', async () => {
    const store = mockStore(initialState);
    await store.dispatch(runGem2s(experimentId));

    expect(fetchMock).toHaveBeenCalled();
  });
});
