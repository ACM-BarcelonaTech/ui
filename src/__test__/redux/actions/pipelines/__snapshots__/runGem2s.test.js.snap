// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runGem2s action Dispatches events properly 1`] = `
Array [
  Object {
    "payload": Object {
      "experimentId": "experiment-id",
    },
    "type": "experimentSettings/backendStatusLoading",
  },
  Object {
    "payload": Object {},
    "type": "experimentSettings/startPipeline",
  },
  Object {
    "payload": Object {
      "experimentId": "experiment-id",
      "experimentName": "Mock experiment",
      "projectId": "project-id",
      "sampleIds": Array [
        "sample-1",
        "sample-2",
      ],
    },
    "type": "experimentSettings/updateInfo",
  },
]
`;

exports[`runGem2s action Dispatches status error if loading fails 1`] = `Array []`;
